src/game/mod.rs

```diff
@@ -258,6 +260,76 @@ impl Game {
         Ok((num_cards, drew_monad))
     }
 
+    pub fn check_player_end(&self, player: usize) -> bool {
+        use self::card::Value::*;
+        let player = &self.players[player];
+        if ! (self.table.common.is_empty() || self.table.discard.is_empty() )  {
+            return false;
+        }
+
+        //-------------------------------------------------------------------
+
+        let mut trade_values: Vec<Option<Value>> =
+            (0..player.hand.len())
+                .combinations(2)
+                .filter_map(|pair| player.trade_value(pair[0], pair[1]).ok())
+                .map(|value| value.succ())
+                .collect();
+
+        trade_values.dedup();
+
+        for value in &trade_values {
+            match value {
+                Some(v) => {
+                    if ! self.table.deck(*v).is_empty() {
+                        return false;
+                    }
+                },
+                None => return false
+            }
+        }
+        drop(trade_values);
+
+        //-------------------------------------------------------------------
+
+        let mut sorted_hand = player.hand.clone();
+        let mut sum = 0;
+        let mut highest_value = Common;
+        let mut num_commons = 0;
+
+        sorted_hand.sort_by_key(|card| card.value);
+        for card in &sorted_hand {
+            if card.is_common() {
+                num_commons += 1;
+            }
+
+            sum += card.num();
+            if card.value > highest_value {
+                highest_value = card.value;
+            }
+            if let Some(value) = highest_value.succ() {
+                if sum >= value.points() {
+                    return false;
+                }
+            }
```

Comment:
This logic still has a problem. If you have enough commons you can buy a card 2 steps up, and this doesn't check for that.